src/glob.zig

```diff
@@ -0,0 +1,2555 @@
+    /// separator:
+    /// `src/**/*.ts` -> `src`, `**`, `*.ts`
+    const Component = struct {
+        start: u32,
+        // length in codepoints
+        len: u32,
+        syntax_hint: SyntaxHint = .None,
+        is_ascii: bool = false,
+        unicode: ?[]u32 = null,
+        const SyntaxHint = enum {
+            None,
+            Single,
+            Double,
+            // Uses special fast-path matching for components like: `*.ts`
+            WildcardFilepath,
+            Literal,
+        };
+    };
+
+    pub fn init(
+        this: *GlobWalker,
+        arena_: Arena,
+        pattern: []const u8,
+        dot: bool,
+        absolute: bool,
+    ) !Maybe(void) {
+        var arena = arena_;
+        errdefer arena.deinit();
+        var cwd: BunString = undefined;
+        switch (Syscall.getcwd(&this.pathBuf)) {
+            .err => |err| {
+                return .{ .err = err };
+            },
+            .result => |result| {
+                var copiedCwd = try arena.allocator().alloc(u8, result.len);
+                @memcpy(copiedCwd, result);
+                cwd = BunString.fromBytes(copiedCwd);
+            },
+        }
+        const globWalker = try this.initWithCwd(arena, pattern, cwd, dot, absolute);
+        return .{ .result = globWalker };
+    }
+
+    pub fn convertUtf8ToCodepoints(codepoints: []u32, pattern: []const u8) void {
+        switch (comptime @import("builtin").target.cpu.arch.endian()) {
+            .Big => {
+                _ = bun.simdutf.convert.utf8.to.utf32.be(pattern, codepoints);
+            },
+            .Little => {
+                _ = bun.simdutf.convert.utf8.to.utf32.le(pattern, codepoints);
+            },
+        }
+    }
+
+    /// `cwd` should be allocated with the arena
+    pub fn initWithCwd(
+        this: *GlobWalker,
+        arena_: Arena,
+        pattern: []const u8,
+        cwd: BunString,
+        dot: bool,
+        absolute: bool,
+    ) !void {
+        var arena = arena_;
+
+        var patternComponents = ArrayList(Component){};
+        try GlobWalker.buildPatternComponents(&arena, &patternComponents, pattern, &this.cp_len, &this.scratch_pattern_codepoints);
+
+        this.patternComponents = patternComponents;
+        this.pattern = pattern;
+        this.arena = arena;
+        this.cwd = cwd;
+        this.dot = dot;
+        this.absolute = absolute;
+    }
+
+    pub fn deinit(this: *GlobWalker) void {
+        this.arena.deinit();
+    }
+
+    pub fn handleSysErrWithPath(
+        this: *GlobWalker,
+        err: Syscall.Error,
+        path_buf: [:0]const u8,
+    ) Syscall.Error {
+        @memcpy(this.pathBuf[0 .. path_buf.len + 1], @as([]const u8, @ptrCast(path_buf[0 .. path_buf.len + 1])));
+        return err.withPath(this.pathBuf[0 .. path_buf.len + 1]);
+    }
+
+    pub fn walk(this: *GlobWalker) !Maybe(void) {
+        if (this.patternComponents.items.len == 0) return .{ .result = undefined };
```

Comment:
Why are we returning undefined? Lets use an empty string or something which indicates it is empty instead of undefined. Undefined is undefined behavior in release builds

Reply:
I changed it to `u0` is that okay?